text/0000-bee-pow/0000-bee-pow.md

```diff
@@ -0,0 +1,332 @@
++ Feature name: `bee-pow` 
++ Start date: 2019-10-03
++ RFC PR: [iotaledger/bee-rfcs#0000](https://github.com/iotaledger/bee-rfcs/pull/0000) 
++ Bee issue: [iotaledger/bee#0000](https://github.com/iotaledger/bee/issues/0000)
+
+# Summary
+
+IOTA is a permissionless and feeless network which unavoidably makes it vulnerable to spam and sybil attacks. To make such
+attacks more costly this RFC proposes a spam protection mechanism based on Proof-of-Work (PoW) where only those
+messages are accepted and propagated by the honest nodes that include a valid nonce value. That nonce serves as a
+proof to show that a predefined amount of computational work has been done upfront either by the publisher itself or
+by some PoW service provider on its behalf.
+
+# Motivation
+
+Since IOTA is a permissionless and feeless network it needs methods of protecting itself against spam from
+intentionally or even unintentionally misbehaving network participants (nodes or clients) that otherwise could
+disrupt the network rather cheaply. Furthermore, this mechanism is currently (at the time of this writing) used in
+the IOTA mainnet, and necessary to build compatible nodes with this framework.
+
+### Job stories:
+
+When I - as a client - want to send a message, for example a value transaction, I want to be able to find and include
+a valid nonce, so that my transaction is accepted and propagated by all honest nodes in the network.
+
+When I - as a node - receive a network message from the gossip pipeline, I want to be able to check if its contained
+nonce value satisfies the pre-defined network difficulty, so that I can decide whether to discard or process that
+message.
+
+# Detailed design
+
+## Hasher
+A hash function always operates on some kind of input data, usually but not necessarily on some kind of `u8` array - and deterministically produces another piece of data - the hash - for example a `u64` that usually takes much less space, and therefore can be used as a label for the data it was derived from. In Rust such a hash function can be generalized using the following trait:
+
+```Rust
+pub trait Hasher {
+    type Data;
+    type Hash;
+
+    fn hash(&mut self, data: &Self::Data) -> Self::Hash;
+}
+```
+
+Notive that by making `Data` and `Hash` associated types this RFC doesn't make any commitment to the underlying representation which allows for a very flexible design.
+
+## Sampler
+
+The `Sampler` trait allows implementing different types that mutate the input data handed to the hasher on each try. The standard sampler used in IOTA simply starts with a zero nonce and increments it until it finds a valid one. There are, however, other samplers possible, like a random sampler.
+
+```Rust
+pub trait Sampler {
+    type Data;
+
+    fn next(&mut self, data: &mut Self::Data);
```

Comment:
`next` should have `bool` as return type and return `false` in case the search data is exhausted. In real world cases we'll hardly exhaust it but it might be useful in tests and as additional guarantees that PoW doesn't loop infinitely.

Reply:
I agree with you.